pkg/controller/hibernation/aws_actuator.go

```diff
@@ -0,0 +1,148 @@
+	logger = logger.WithField("cloud", "aws")
+	logger.Info("stopping machines")
+	awsClient, err := a.awsClientFn(logger, cd, c)
+	if err != nil {
+		return err
+	}
+	instanceIDs, err := getClusterInstanceIDs(logger, cd, awsClient, []string{"pending", "running"})
+	if err != nil {
+		return err
+	}
+	if len(instanceIDs) == 0 {
+		logger.Warning("no instances were found to stop")
+		return nil
+	}
+	_, err = awsClient.StopInstances(&ec2.StopInstancesInput{
+		InstanceIds: instanceIDs,
+	})
+	if err != nil {
+		logger.WithError(err).Error("failed to stop instances")
+	}
+	return err
+}
+
+// StartMachines will select machines belonging to the given ClusterDeployment
+func (a *awsActuator) StartMachines(logger log.FieldLogger, cd *hivev1.ClusterDeployment, c client.Client) error {
+	logger = logger.WithField("cloud", "aws")
+	logger.Infof("starting machines")
+	awsClient, err := a.awsClientFn(logger, cd, c)
+	if err != nil {
+		return err
+	}
+	instanceIDs, err := getClusterInstanceIDs(logger, cd, awsClient, []string{"stopping", "stopped", "shutting-down"})
+	if err != nil {
+		return err
+	}
+	if len(instanceIDs) == 0 {
+		logger.Warning("no instances were found to start")
+		return nil
+	}
+	_, err = awsClient.StartInstances(&ec2.StartInstancesInput{
+		InstanceIds: instanceIDs,
+	})
+	if err != nil {
+		logger.WithError(err).Error("failed to start instances")
+	}
+	return err
+}
+
+// MachinesRunning will return true if the machines associated with the given
+// ClusterDeployment are in a running state.
+func (a *awsActuator) MachinesRunning(logger log.FieldLogger, cd *hivev1.ClusterDeployment, c client.Client) (bool, error) {
+	logger = logger.WithField("cloud", "aws")
+	logger.Infof("checking whether machines are running")
+	awsClient, err := a.awsClientFn(logger, cd, c)
+	if err != nil {
+		return false, err
+	}
+	instanceIDs, err := getClusterInstanceIDs(logger, cd, awsClient, []string{"stopping", "stopped", "pending", "shutting-down"})
+	if err != nil {
+		return false, err
+	}
+	return len(instanceIDs) == 0, nil
+}
+
+// MachinesStopped will return true if the machines associated with the given
+// ClusterDeployment are in a stopped state.
+func (a *awsActuator) MachinesStopped(logger log.FieldLogger, cd *hivev1.ClusterDeployment, c client.Client) (bool, error) {
+	logger = logger.WithField("cloud", "aws")
+	logger.Infof("checking whether machines are stopped")
+	awsClient, err := a.awsClientFn(logger, cd, c)
+	if err != nil {
+		return false, err
+	}
+	instanceIDs, err := getClusterInstanceIDs(logger, cd, awsClient, []string{"running", "pending", "stopping", "shutting-down"})
+	if err != nil {
+		return false, err
+	}
+	return len(instanceIDs) == 0, nil
+}
+
+func getAWSClient(logger log.FieldLogger, cd *hivev1.ClusterDeployment, c client.Client) (awsclient.Client, error) {
+	awsClient, err := awsclient.NewClient(c, cd.Spec.Platform.AWS.CredentialsSecretRef.Name, cd.Namespace, cd.Spec.Platform.AWS.Region)
```

Comment:
Are we losing any efficiency with the cloud-provider calls by creating a new client for each request made, as opposed to using a single client for the entire reconciliation?

Reply:
Looking at the controller logic it looks like we only make one actuator call per reconciliation, so I don't think we would gain much.

Reply:
:+1: